prov/rxm/src/rxm_conn.c

```diff
@@ -136,59 +157,60 @@ static void rxm_send_queue_close(struct rxm_send_queue *send_queue)
 static void rxm_conn_close(struct util_cmap_handle *handle)
 {
 	struct rxm_conn *rxm_conn = container_of(handle, struct rxm_conn, handle);
-
+	struct rxm_ep *rxm_ep = container_of(handle->cmap->ep, struct rxm_ep, util_ep);
 	if (!rxm_conn->msg_ep)
 		return;
-	/* Save the preposted RX list for further cleanup */
-	dlist_splice_tail(&rxm_conn->saved_posted_rx_list,
-			  &rxm_conn->posted_rx_list);
-	rxm_conn->saved_msg_ep = rxm_conn->msg_ep;
-	FI_DBG(&rxm_prov, FI_LOG_EP_CTRL,
-	       "Saved MSG EP fid for further deletion in main thread\n");
-	rxm_conn->msg_ep = NULL;
+	struct rxm_cmd_conn_free_data conn_free_data = {
+		.type = RXM_CMD_CONN_CLOSE,
+		.rxm_conn = rxm_conn,
+	};
+	struct util_cmap_cmd *cmd =
+		calloc(1, sizeof(struct util_cmap_cmd) +
+			  sizeof(struct rxm_cmd_conn_free_data));
+	if (!cmd) {
+		rxm_conn_close_msg_ep(rxm_ep, rxm_conn);
+		return;
+	}
+	cmd->type = UTIL_CMAP_CMD_INTERNAL;
+	memcpy(cmd->data, &conn_free_data, sizeof(conn_free_data));
+	ofi_queue_push(&cmd->entry, rxm_ep->util_ep.cmap->cmd_queue);
+	rxm_ep->util_ep.cmap->cmd_write++;
+	rxm_conn_wake_up_wait_obj(rxm_ep);
 }
 
 static void rxm_conn_free(struct util_cmap_handle *handle)
 {
 	struct rxm_ep *rxm_ep = container_of(handle->cmap->ep, struct rxm_ep, util_ep);
 	struct rxm_conn *rxm_conn = container_of(handle, struct rxm_conn, handle);
 
-	/* This handles case when saved_msg_ep wasn't closed */
-	if (rxm_conn->saved_msg_ep) {
-		rxm_ep_cleanup_posted_rx_list(rxm_ep, &rxm_conn->saved_posted_rx_list);
-		if (fi_close(&rxm_conn->saved_msg_ep->fid))
-			FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
-				"Unable to close saved msg_ep\n");
+	if (handle->cmap == UTIL_CM_THREAD_OPEN) {
```

Comment:
Should this be `handle->cmap->cm_thread_state`?

Reply:
Good catch. Thanks!